src/SAWScript/Proof.hs

```diff
@@ -5,24 +5,396 @@ License     : BSD3
+    --   allowed to appear inside the evidence as a local assumption.
+  | AssumeEvidence Prop Evidence
+
+    -- | This type of evidence is used to prove a universally-quantified statement.
+  | ForallEvidence (ExtCns Term) Evidence
+
+    -- | This type of evidence is used to weaken a goal by adding a hypothesis,
+    --   where the hypothesis is proved by the given theorem.
+  | CutEvidence Theorem Evidence
+
+    -- | This type of evidence is used to modify a goal to prove via rewriting.
+    --   The goal to prove is rewriten by the given simpset; then the provided
+    --   evidence is used to check the modified goal.
+  | RewriteEvidence Simpset Evidence
+
+    -- | This type of evidence is used to modify a goal to prove via unfolding
+    --   constant definitions.  The goal to prove is modified by unfolding
+    --   constants identified via the given set of @VarIndex@; then the provided
+    --   evidence is used to check the modified goal.
+  | UnfoldEvidence (Set VarIndex) Evidence
+
+    -- | This type of evidence is used to modify a goal to prove via evaluation
+    --   into the the What4 formula representation. During evaluation, the
+    --   constants identified by the given set of @VarIndex@ are held
+    --   uninterpreted (i.e., will not be unfolded).  Then, the provided
+    --   evidence is use to check the modified goal.
+  | EvalEvidence (Set VarIndex) Evidence
+
+-- | The the proposition proved by a given theorem.
+thmProp :: Theorem -> Prop
+thmProp (LocalAssumption p) = p
+thmProp Theorem{ _thmProp = p } = p
+
+-- | Retrieve any solver stats from the proved theorem.
+thmStats :: Theorem -> SolverStats
+thmStats (LocalAssumption _) = mempty
+thmStats Theorem{ _thmStats = stats } = stats
+
+-- | Retrive the evidence associated with this theorem.
+thmEvidence :: Theorem -> Evidence
+thmEvidence (LocalAssumption p) = LocalAssumptionEvidence p
+thmEvidence Theorem{ _thmEvidence = e } = e
+
+impossibleEvidence :: [Evidence] -> IO Evidence
+impossibleEvidence _ = fail "impossibleEvidence: attempted to check an impossible proof!"
+
+splitEvidence :: [Evidence] -> IO Evidence
+splitEvidence [e1,e2] = pure (SplitEvidence e1 e2)
+splitEvidence _ = fail "splitEvidence: expected two evidence values"
+
+assumeEvidence :: Prop -> [Evidence] -> IO Evidence
+assumeEvidence p [e] = pure (AssumeEvidence p e)
+assumeEvidence _ _ = fail "assumeEvidence: expected one evidence value"
+
+forallEvidence :: ExtCns Term -> [Evidence] -> IO Evidence
+forallEvidence x [e] = pure (ForallEvidence x e)
+forallEvidence _ _ = fail "forallEvidence: expected one evidence value"
+
+cutEvidence :: Theorem -> [Evidence] -> IO Evidence
+cutEvidence thm [e] = pure (CutEvidence thm e)
+cutEvidence _ _ = fail "cutEvidence: expected one evidence value"
+
+-- | Construct a theorem directly via a proof term.
+proofByTerm :: SharedContext -> Term -> IO Theorem
+proofByTerm sc prf =
+  do ty <- scTypeOf sc prf
```

Comment:
Should we do the more exhaustive type checking here rather than assuming the term is well-typed? We do the more exhaustive checking at `checkEvidence` time, so it's not clear.

Reply:
I think we should consider well-typedness to be an internal invariant of the `Term` type; in the long term we should be enforcing typing rules as `Term`s are built. So I think it makes sense to assume that it's well-formed here.

Reply:
Makes sense to me.